### ISSUE TITLE
Switch build failing (broken compiler options)

### ISSUE THREAD

**Issue by AJenbo** [diasurgical/devilutionX#763]
https://circleci.com/gh/diasurgical/devilutionX/11688
`aarch64-none-elf-gcc: fatal error: /opt/devkitpro/libnx/switch.specs: attempt to rename spec 'link' to already defined spec 'old_link'`

Removing `set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -specs=${DEVKITPRO}/libnx/switch.specs")` from `/opt/devkitpro/switch.cmake` appears to solve the first issue, but then we run in to linker issues.

```
/opt/devkitpro/devkitA64/lib/gcc/aarch64-none-elf/10.1.0/../../../../aarch64-none-elf/bin/ld: /opt/devkitpro/libnx/lib/libnx.a(switch_crt0.o): in function `_start':
/home/fincs/pacman-packages/libnx/src/libnx-3.3.0/nx/source/runtime/switch_crt0.s:5: multiple definition of `_start'; /opt/devkitpro/devkitA64/lib/gcc/aarch64-none-elf/10.1.0/../../../../aarch64-none-elf/lib/crt0.o:(.text+0x0): first defined here
/opt/devkitpro/devkitA64/lib/gcc/aarch64-none-elf/10.1.0/../../../../aarch64-none-elf/bin/ld: cannot find -lfreetype
```

Last known good build was:
https://circleci.com/gh/diasurgical/devilutionX/11627

**Comment by AJenbo** [comment]
@rsn8887 do you know of any solutions to this?

**Comment by AJenbo** [comment]
Probably related to this change: https://github.com/devkitPro/pacman-packages/commit/bf656bf1e992fd9f26403d126da14bc03756a18c#diff-c9c6bfc70698b46d91f4f7bdafb9aab4

**Comment by rsn8887** [comment]
Maybe only include the switch.cmake and don’t use the toolchain file anymore?

**Comment by AJenbo** [comment]
@WinterMute there is nothing we would like more than to help get this in a better state. I'm not sure what/who you are quoting, it's not from this issue and I can't find it even when searching on Google. So you are not exactly being helpful either.

We haven't reported this anywhere yet since we are still investigating the issue to determine what the cause is. Also, we generally feel that we are meet with hostility if we report issues to the DevkitPro-project, especially if they contain incorrect assumptions, mistakes on our part, or are in the wrong section of the project. And now even if we start to figure out if we did something wrong ourselves first.

> We want toolchain files that work so people aren't adding their own odd twist and copying the same semi-broken and fragile things around projects.

Are you asking for us to provide you with this? We are more than happy to help test things and collaborate. We are not experts on CMake, toolchains, or Nintendo Switch (non of the maintainers even own one). We have only added what we have out of necessity and best-effort. There is nothing we would rather do than rely fully on a toolchain maintained config.

Implying that what we have is "semi broken and fragile" isn't really adding anything to the conversation, we already know this - hens this issue. It's probably also true that our solution "have some fundamental errors", but this doesn't help us understand how to improve it, and it sure is discouraging to the though that you might be interested in anything from our side.

In short if deleting thel CMake/switch folder meant that the project would build for Switch we would do it in a heartbeat, please help.

**Comment by WinterMute** [comment]
> @WinterMute I'm not sure what/who you are quoting, it's not from this issue and I can't find it even when searching on Google. So you are not exactly being helpful either.


Check your emails from this issue. I guess @rsn8887 deleted his comment after I said it was unhelpful. It was presumably a reply to https://github.com/diasurgical/devilutionX/issues/763#issuecomment-662366661

> There is nothing we would like more than to help get things in a better state. We haven't reported this anywhere yet since we are still investigating the issue to determine what the cause is. Also, we generally feel that we are met with hostility if we report issues to the DevkitPro-project, especially if they contain incorrect assumptions, mistakes on our part, or are in the wrong section of the project. And now even if we start to figure out if we did something wrong ourselves first.

I don't really know what you're referring to re: reporting issues and the (now deleted) comment I quoted was suggesting a "fix". 

Really letting us know about this would have been a good first step. We know the cmake stuff is broken and we want to sort it out so this kind of issue stops happening. My concern is that people just work around it & then the workaround propagates until the next time I have a go at fixing cmake for something else when it all comes crumbling down again.

> > We want toolchain files that work so people aren't adding their own odd twist and copying the same semi-broken and fragile things around projects.
> 
> Are you asking for us to provide you with this? We are more than happy to help test things and collaborate. We are not experts on CMake, toolchains, or Nintendo Switch (non of the maintainers even own one). We have only added what we have out of necessity and best-effort. There is nothing we would rather do than rely fully on a toolchain maintained config.
> 
> Implying that what we have is "semi broken and fragile" isn't really adding anything to the conversation, we already know this - hens this issue. It's probably also true that our solution "has some fundamental errors", but this doesn't help us understand how to improve it, and it sure is discouraging to the thought that you might be interested in anything from our side.

It's not your fault it's broken and fragile, don't take it personally. The switch related cmake files were copied from a template project put together without anyone bothering to talk to us first. I'd prefer if that template were expunged from the internet so it no longer ensnares the unwary in a pile of technical debt but unfortunately I can't have that. If people are going to use cmake then the support framework for it needs to be supplied with our tools, not copied & pasted into every project under the sun.

fwiw, I include the devkitPro supplied cmake stuff in the broken and fragile territory. The general feeling I get is that nobody really quite knows how cmake is supposed to work especially when it comes to cross compiling and most especially when it comes to the particular kind of cross compiling I've specialised in. It's deeply frustrating for everyone.

> In short if deleting the CMake/switch folder meant that the project would build for Switch we would do it in a heartbeat, but that doesn't appear to be the case. please help.

We'll get there eventually. Assuming people communicate.

fwiw @rsn8887 has private access to us atm. We can always get a message that way. It's probably kind of preferable that we talk with people actually working with switch code from the point of view of actually being able to test the switch binaries.

**Comment by AJenbo** [comment]
@WinterMute Ok, that explains it :) I have disabled notifications via email as I check GitHub quite often and get a high number of notifications that I don't want in my email inbox.

Thanks for clarifying you view on things. I will try to notify you early on when future issues happen.
 
Sorry for the late response to your comment but I had to take care of some other things in life for a bit.

In a cruel twist of faith, it appears that builds have started falling again same day as we merged @rsn8887's solution. Now due to some Docker related error, the following is the log from the build on the 1st of August, do you have any idea what might be causing this?

```
Build-agent version 1.0.37730-128dacb8 (2020-07-30T15:03:44+0000)
Docker Engine Version: 18.09.6
Kernel Version: Linux f11da0b945a6 4.15.0-1077-aws #81-Ubuntu SMP Wed Jun 24 16:48:15 UTC 2020 x86_64 Linux
Starting container devkitpro/devkita64:latest
  image cache not found on this host, downloading devkitpro/devkita64:latest
latest: Pulling from devkitpro/devkita64

77b12acb: Already exists 
46bc7d40: Pulling fs layer 
2ab5cd05: Pulling fs layer 
df2f57d3: Extracting [==================================================>]  256.2MB/256.2MBB

CircleCI was unable to start the container because of a userns remapping failure in Docker.

This typically means the image contains files with UID/GID values that are higher than Docker and CircleCI can use.

Checkout our docs https://circleci.com/docs/2.0/high-uid-error/ to learn how to fix this problem.
Original error: CircleCI was unable to start the container because of a userns remapping failure in Docker.

This typically means the image contains files with UID/GID values that are higher than Docker and CircleCI can use.

Checkout our docs https://circleci.com/docs/2.0/high-uid-error/ to learn how to fix this problem.
Original error: failed to register layer: Error processing tar file(exit status 1): Container ID 197609 cannot be mapped to a host ID
```
